packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
   private validateArtifacts(): string[] {
     const ret = new Array<string>();
 
-    const outputArtifactNames = new Set<string>();
-    for (const stage of this._stages) {
-      const sortedActions = stage.actionDescriptors.sort((a1, a2) => a1.runOrder - a2.runOrder);
-
-      for (const action of sortedActions) {
-        // start with inputs
-        const inputArtifacts = action.inputs;
-        for (const inputArtifact of inputArtifacts) {
-          if (!inputArtifact.artifactName) {
-            ret.push(`Action '${action.actionName}' has an unnamed input Artifact that's not used as an output`);
-          } else if (!outputArtifactNames.has(inputArtifact.artifactName)) {
-            ret.push(`Artifact '${inputArtifact.artifactName}' was used as input before being used as output`);
+    interface PipelineLocation {
+      stageIndex: number;
+      stageName: string;
+      action: FullActionDescriptor;
+    }
+
+    /**
+     * Returns whether a is before or the same order as b
+     */
+    function beforeOrEqual(a: PipelineLocation, b: PipelineLocation) {
+      if (a.stageIndex !== b.stageIndex) { return a.stageIndex < b.stageIndex; }
+      return a.action.runOrder <= b.action.runOrder;
+    }
+
+    function renderLoc(loc: PipelineLocation) {
+      // runOrders are 1-based, so make the stageIndex also 1-based otherwise it's going to be confusing.
+      return `stage ${loc.stageIndex + 1} ('${loc.stageName}') action ${loc.action.runOrder} ('${loc.action.actionName}')`;
+    }
+
+    const producer: Record<string, PipelineLocation> = {};
+    const firstConsumer: Record<string, PipelineLocation> = {};
```

Comment:
Can you change these to plurals? So, `producers` and `firstConsumers`. It's actually pretty confusing to have `producer` and `firstConsumer` be collections, but `thisLocation` and `producerLoc` below as single values.

Reply:
But... but... but... what about the beautiful duality between a function and a map?

If there had been a function which would give you the producer location from an artifact ID, would you have given it the plural name?

```ts
// This is a function that gets me the value I want
const loc = producer('abc')

// This is a map that gets me the value I want
const loc = producer['abc']

// -----------------------------
// Sooo... you would prefer this?
const loc = producers['abc']

// Would you also write this?
const loc = producers('abc');

// ...and if so, why not? 😛 
```

The only difference is that the value from the map is precomputed, and the value from the function is... well... maybe it's precomputed, maybe it's not, no way to tell. Why would you even care?

Why the preferred focus of the map being a literal bag of values, as opposed to a mapping that can get you the value that you want?